pkg/sources/registration.go

```
@@ -0,0 +1,195 @@
+	mappedResources []string
+	mappedEvent     config.EventType
+}
+
+func (i Registration) handleEvent(ctx context.Context, resource string, target config.EventType, sourceRoutes []Route, fn eventHandler) {
+	switch target {
+	case config.CreateEvent:
+		i.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				sources := sourcesForObjNamespace(ctx, sourceRoutes, obj, i.log, i.mapper, i.dynamicCli)
+				i.log.Debugf("handle CreateEvent, resource: %s, sources: %+v", resource, sources)
+				if len(sources) > 0 {
+					fn(ctx, resource, sources, nil)(obj)
+				}
+			},
+		})
+	case config.DeleteEvent:
+		i.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+			DeleteFunc: func(obj interface{}) {
+				sources := sourcesForObjNamespace(ctx, sourceRoutes, obj, i.log, i.mapper, i.dynamicCli)
+				i.log.Debugf("handle DeleteEvent, resource: %s, sources: %+v", resource, sources)
+				if len(sources) > 0 {
+					fn(ctx, resource, sources, nil)(obj)
+				}
+			},
+		})
+	case config.UpdateEvent:
+		i.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				sources, diffs := qualifySourcesForUpdate(ctx, newObj, oldObj, sourceRoutes, i.log, i.mapper, i.dynamicCli)
+				i.log.Debugf("handle UpdateEvent, resource: %s, sources: %+v, diffs: %+v", resource, sources, diffs)
+				if len(sources) > 0 {
+					fn(ctx, resource, sources, diffs)(newObj)
+				}
+			},
+		})
+	}
+}
+
+func (i Registration) handleMapped(ctx context.Context, targetEvent config.EventType, routeTable map[string][]Entry, fn eventHandler) {
+	i.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			var eventObj coreV1.Event
+			err := utils.TransformIntoTypedObject(obj.(*unstructured.Unstructured), &eventObj)
+			if err != nil {
+				i.log.Errorf("Unable to transform object type: %v, into type: %v", reflect.TypeOf(obj), reflect.TypeOf(eventObj))
```

Any reason for swallowing the original error here? Passing the returned error messages here would be valuable for better error chain visibility. Especially good for concurrent calls.

---

No reason at all.

The code was moved from another section in `pkg/controller/controller.go`. When error returns fit they will be returned 😄.

---

Actually after digging in, notice that the container func `AddFunc: func(obj interface{}) {` defines no error returns.

Meaning, the error cannot be returned.